### ISSUE TITLE
Error locating diff-highlight

### ISSUE THREAD

**Issue by giggio** [so-fancy/diff-so-fancy#238]
On Windows, when I run:

````powershell
git diff --color | diff-so-fancy
````

I get:

````
Error locating diff-highlight
````

Running `diff-so-fancy --help` displays the help, and the output of `diff-so-fancy --version` is:

````
Diff-so-fancy: https://github.com/so-fancy/diff-so-fancy
Version      : 1.0.0
````

I can find diff-hightlight on the path:
![image](https://user-images.githubusercontent.com/334958/27316490-1d1204cc-554e-11e7-817e-105fed0cbc8c.png)

If you need more info, let me know.

**Comment by OJFord** [comment]
Have you just started using dsf, or did this start happening when you updated to v1.0.0?

**Comment by giggio** [comment]
I was on version 0.11.4, and it worked fine.

**Comment by scottchiefbaker** [comment]
This does not surprise me... we changed how we detect/run `diff-highlight` in 1.0.0. I don't have a way to test on Windows, so I'll need some help on this issue. Specificially the code is here:

https://github.com/so-fancy/diff-so-fancy/blob/master/diff-so-fancy#L574

We rely on `which` but that no might be available/reliable on Windows? I wonder if the issue (that you point out above) is that `diff-highlight` has a `.CMD` on the end? Or is that a Windows way of saying it's an executable? Basically we need to make sure that `find-diff-highlight` function returns a valid path on Windows.

I'm open to any feedback on this as I don't know Windows shell enough to know which direction to go.

**Comment by scottchiefbaker** [comment]
@sudo-suhas since you use Windows, can you offer any feedback on this? I'm not sure what the next troubleshooting steps are.

**Comment by sudo-suhas** [comment]
@scottchiefbaker I'll try to replicate this tomorrow. 

To be clear, the package isn't always installed using npm right? Otherwise we could use some node specific, platform independent method to find the executable.

**Comment by scottchiefbaker** [comment]
Node is not a requirement... we're working on an installation script that would simplify all this: #218

**Comment by giggio** [comment]
Adding more info:

DSF will not work on Windows if perl is not available. It is available to me as I add the `bin` dir from git to the path (in my case `C:\Program Files\Git\usr\bin\`).
For example, `which` for me, comes from `C:\Program Files\Git\usr\bin\which.exe`, running in PowerShell `Get-Command which`, I get:

````
CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Application     which.exe                                          0.0.0.0    C:\Program Files\Git\usr\bin\which.exe
````

And `which which` will show `/usr/bin/which`.

If I do `git diff | diff-so-fancy`, this is the error I get:

````
Can't exec "/c/Program": No such file or directory at C:\Program Files\nodejs\node_modules\diff-so-fancy\diff-so-fancy line 568.
````

Line 568 shows (https://github.com/so-fancy/diff-so-fancy/blob/v1.0.0/diff-so-fancy#L568):

````perl
my $pid = open(my $cmd_output, "-|", $dh);
````

Which is just after the `find_diff_highlight` function call.

If I run `which diff-highlight` I get:

````
/c/Program Files/nodejs/diff-highlight
````

I am assuming that what is happening is that perl is getting `/c/Program Files/nodejs/diff-highlight` as the return value from `which diff-hightlight`.

And when perl tries to open that path, it fails. Maybe because of the space in the middle of the path, or maybe for some other reason.

**Comment by scottchiefbaker** [comment]
This may have been fixed TODAY in #240. That was a Windows specific bug AFAIK. Can you test again?

**Comment by giggio** [comment]
It is also interesting to point out that I also use the npm module `which`, which then errors with `Error location diff-hightlight`, which I pointed out when I opened the issue.
So, clearly, there is a problem of using `which` on Windows. If it is the cywin/msysgit version, it will output something, if it is the npm module, something else.
Here is the output from the npm which module:
````
C:\Program Files\nodejs\diff-highlight.CMD
````

Which is differerent from the git bash one I showed on the previous comment.

**Comment by giggio** [comment]
@scottchiefbaker I just checked and it works if I don't use the which module from npm.

**Comment by scottchiefbaker** [comment]
Just to clarify:

npm `which` does **not** work  
cygwin `which` does work  

??

@giggio how do you specify which one you want to run? Can we modify d-s-f to specifically use the cygwin one?

**Comment by giggio** [comment]
I am not sure if npm or cygwin is more popular. But I guess you may get bug reports if you support one and not the other. I am fine not using cygwin's.

And yes, you understanding is right, npm `which` does not work, cygwin `which` works.

BTW, as the fix is out, would you please publish it to npm so I can go back to having beautiful diffs? :)

**Comment by sudo-suhas** [comment]
> would you please publish it to npm

@scottchiefbaker Please do publish the patched version.

Like @giggio mentioned, the output of git `which` is different from the node.js one.
```
λ where which
C:\Program Files\Git\usr\bin\which.exe
C:\Program Files\nodejs\which
C:\Program Files\nodejs\which.cmd

# git which
λ which diff-highlight
/c/Program Files/nodejs/diff-highlight

λ "c:\Program Files\nodejs\which" diff-highlight
C:\Program Files\nodejs\diff-highlight.CMD
```

I don't think we can be sure about which executable will be picked up which `diff-so-fancy` tries to locate `diff-highlight`. Because on my system, although I installed `which` using npm, it still works fine. 

I think this issue can be fixed with [`cygpath`](https://superuser.com/questions/1025680/how-to-convert-a-windows-path-with-backslashes-to-unix-format-on-babun-cygwin).
```
λ cygpath "C:\Program Files\nodejs\diff-highlight.CMD"
/c/Program Files/nodejs/diff-highlight.CMD
```
Shall I make a pull request for this?

**Comment by scottchiefbaker** [comment]
Can you try my experimental branch here:

https://github.com/scottchiefbaker/diff-so-fancy/blob/scott_which/diff-so-fancy

I removed the requirement to shell out and run `which`.

**Comment by giggio** [comment]
@scottchiefbaker I get `Error #19391: diff-highlight (1) not executable` on Windows with your branch.

**Comment by scottchiefbaker** [comment]
@sudo-suhas We can't require cygwin... d-s-f _almost_ runs on vanilla Windows + Perl. I'm wondering if perhaps we just do a check and if you're on Windows it calls `where` instead of `which`. 

@giggio can you modify line 576 to look something like: 

    my $dh_in_path = trim(`where $dh`);

and test that instead?

**Comment by giggio** [comment]
@scottchiefbaker it works with this line changed.

**Comment by scottchiefbaker** [comment]
Ok let's go with that for now... would you like to do a pull request that tests for:

```
$IS_WIN => ($^O eq 'MSWin32' or $^O eq 'dos' or $^O eq 'os2');
$IS_CYG => ($^O eq 'cygwin');

if ($IS_WIN || $IS_CYG) {
    # Use where 
} else {
    # Use which
}
```

**Comment by giggio** [comment]
I don't really fell comfortable with Perl to work on a PR.

**Comment by scottchiefbaker** [comment]
@giggio fair enough :)

Can you try this test branch which implements it that way: https://github.com/scottchiefbaker/diff-so-fancy/blob/scott_which/diff-so-fancy

**Comment by giggio** [comment]
I just pulled your code.
If I call it from PowerShell or cmd, it works. But if I install npm's `which`, it fails.
If I call it from git bash, it always fails.

As I said, I don't know any Perl, but I tried opening a repl as stated [here](https://stackoverflow.com/questions/73667/how-can-i-start-an-interactive-console-for-perl), and when wrote `print $^O` it answered with `msys`. Maybe you need to include that? I made line 580 like this, and then all worked:

````perl
if ($^O eq 'MSWin32' or $^O eq 'dos' or $^O eq 'os2' or $^O eq 'cygwin' or $^O eq 'msys')
````

I hope it helps. :)

**Comment by scottchiefbaker** [comment]
Actually that is very helpful. Can you be 100% clear what your Window environment is?

Windows 10 with vanilla `cmd` (not cygwin and not Bash)? Sounds like we just need to expand our `is_windows()` check.

**Comment by giggio** [comment]
Latest Windows 10. Perl does not come with Windows, I am using the one that comes with Git.

This is what I get from `perl --version`:

````
perl --version

This is perl 5, version 24, subversion 1 (v5.24.1) built for x86_64-msys-thread-multi
````

Perl is in `C:\Program Files\Git\usr\bin\perl.EXE`.

This is all straight from PowerShell or cmd.

**Comment by scottchiefbaker** [comment]
Interesting... so the only thing you installed was Git on Windows? And that includes "git bash" and Perl? We'll have to test with a vanilla setup like that.

When I test the bash shell that comes with Git Perl then reports it as `msys`.

Powershell: `MSWin32`
cmd.exe: `MSWin32`
Git Bash: `msys`

**Comment by giggio** [comment]
Yes, only Git for Windows. Perl comes with it. Git bash as well, but I ran the commands I posted earlier from PowerShell and Cmd.

**Comment by scottchiefbaker** [comment]
Gotcha... do you want to submit a PR for that simple `msys` change on line 580 that fixed things for you?

**Comment by giggio** [comment]
Done!
https://github.com/scottchiefbaker/diff-so-fancy/pull/5

**Comment by jacktho** [comment]
I just ran into this problem as well while trying out dsf. Hopefully the fix can be published to NPM soon.

**Comment by scottchiefbaker** [comment]
The fix is live on the master branch which is 1.0.1 right now. I'm not sure if we're going to do a full-blown NPM release for this or not. For now feel free to just download the d-s-f script and overwrite the one you have in place now.
